tests/components/bsblan/test_config_flow.py

```diff
@@ -0,0 +1,92 @@
+from homeassistant.config_entries import SOURCE_USER
+from homeassistant.const import CONF_HOST, CONF_PORT
+from homeassistant.core import HomeAssistant
+
+from . import init_integration
+
+from tests.common import load_fixture
+from tests.test_util.aiohttp import AiohttpClientMocker
+
+
+async def test_show_user_form(hass: HomeAssistant) -> None:
+    """Test that the user set up form is served."""
+    result = await hass.config_entries.flow.async_init(
+        config_flow.DOMAIN, context={"source": SOURCE_USER},
+    )
+
+    assert result["step_id"] == "user"
+    assert result["type"] == data_entry_flow.RESULT_TYPE_FORM
+
+
+async def test_connection_error(
+    hass: HomeAssistant, aioclient_mock: AiohttpClientMocker
+) -> None:
+    """Test we show user form on BSBLan connection error."""
+    aioclient_mock.post(
+        "http://example.local:80/1234/JQ?Parameter=6224,6225,6226",
+        exc=aiohttp.ClientError,
+    )
+
+    result = await hass.config_entries.flow.async_init(
+        config_flow.DOMAIN,
+        context={"source": SOURCE_USER},
+        data={CONF_HOST: "example.local", CONF_PASSKEY: "1234", CONF_PORT: 80},
+    )
+
+    assert result["errors"] == {"base": "connection_error"}
+    assert result["step_id"] == "user"
+    assert result["type"] == data_entry_flow.RESULT_TYPE_FORM
+
+
+async def test_user_device_exists_abort(
+    hass: HomeAssistant, aioclient_mock: AiohttpClientMocker
+) -> None:
+    """Test we abort zeroconf flow if BSBLan device already configured."""
+    await init_integration(hass, aioclient_mock)
+
+    result = await hass.config_entries.flow.async_init(
+        config_flow.DOMAIN,
+        context={"source": SOURCE_USER},
+        data={CONF_HOST: "example.local", CONF_PASSKEY: "1234", CONF_PORT: 80},
+    )
+
+    assert result["type"] == data_entry_flow.RESULT_TYPE_ABORT
+
+
+async def test_full_user_flow_implementation(
+    hass: HomeAssistant, aioclient_mock
+) -> None:
+    """Test the full manual user flow from start to finish."""
+    aioclient_mock.post(
+        "http://example.local:80/1234/JQ?Parameter=6224,6225,6226",
+        text=load_fixture("bsblan/info.json"),
+        headers={"Content-Type": "application/json"},
+    )
+
+    result = await hass.config_entries.flow.async_init(
+        config_flow.DOMAIN, context={"source": SOURCE_USER},
+    )
+
+    assert result["step_id"] == "user"
+    assert result["type"] == data_entry_flow.RESULT_TYPE_FORM
+
+    result = await hass.config_entries.flow.async_configure(
+        result["flow_id"],
+        user_input={CONF_HOST: "example.local", CONF_PASSKEY: "1234", CONF_PORT: 80},
+    )
+
+    assert result["data"][CONF_HOST] == "example.local"
+    assert result["data"][CONF_PASSKEY] == "1234"
+    assert result["data"][CONF_PORT] == 80
+    assert result["data"][CONF_DEVICE_IDENT] == "RVS21.831F/127"
+    assert result["title"] == "RVS21.831F/127"
```

Comment:
![Codacy](https://app.codacy.com/assets/images/favicon.png) Issue found: [Use of assert detected. The enclosed code will be removed when compiling to optimised byte code.](https://app.codacy.com/manual/foreign-sub/home-assistant/pullRequest?prid=5518609)